### PR TITLE
Various fixes, run tests in GH workflow

### PR DESCRIPTION
**Pull request by martinpitt**
I took my first look at ferny, and couldn't resist..

With these fixes, I can run `pytest` locally successfully.

[test success on my fork](https://github.com/martinpitt/ferny/actions/runs/3930861781)

### CODE REVIEW

**Comment by martinpitt** [comment]
@allisonkarlitskaya For your reviewing pleasure :grin:

**Comment by martinpitt** [comment]
The test fails on Ubuntu 22.04 with

    FAILED test/test_basic.py::TestBasic::test_rsa_key - RuntimeError: b'Host key verification failed.\r\n'

I can fix that with changing the options in the test to

    options={'UserKnownHostsFile': '/dev/null', 'StrictHostKeyChecking': 'no'},

Then it fails differently:

     RuntimeError: b"Warning: Permanently added '[127.0.0.1]:37131' (RSA) to the list of known hosts.\r\nadmin@127.0.0.1: Permission denied (publickey).\r\n"

It passes when I set `identity_file` to `test/id_rsa`, but I figure the point of the exercise is to use the encrypted key, so that SSH actually has a reason to ask for a password. I think this will require some better logging, right now this is too opaque.

File: pyproject.toml [hunk 1]
```diff
@@ -8,6 +8,9 @@ version = "0"
 
 [tool.mypy]
 mypy_path = 'src'
+[[tool.mypy.overrides]]
+ module = ["mockssh"]
```
Comment (allisonkarlitskaya) [inline pyproject.toml:15]: Is this extra space relevant?  I don't do toml.

File: pyproject.toml [hunk 2]
```diff
@@ -6,6 +6,9 @@ build-backend = "setuptools.build_meta"
 name = "ferny"
 version = "0"
 
+[project.optional-dependencies]
+test = ['mock-ssh-server', 'pytest']
```
Comment (allisonkarlitskaya) [inline pyproject.toml:10]: Ugh.  I have to say that I'm kinda not a fan of this.  Is there some standard that says that it ought to be done this way?

In general, I prefer to have less stuff that pip might use as an excuse to randomly download crap from the internet without asking about it...

File: test/test_basic.py [hunk 3]
```diff
@@ -18,6 +18,12 @@ async def askpass(self, msg, hint):
 
 
 class TestBasic(unittest.IsolatedAsyncioTestCase):
+    @classmethod
+    def setUpClass(kls):
+        # git does not track file permissions, and SSH fails on group/world readability
+        os.chmod('test/id_rsa', 0o600)
```
Comment (allisonkarlitskaya) [inline test/test_basic.py:25]: I don't care for this — it's effectively modifying the source tree.  Isn't there a way we can tell SSH to be less strict about this?  If that's the way it's gotta be then we should create temporary copies of these elsewhere.

File: src/ferny.py [hunk 4]
```diff
@@ -29,7 +29,7 @@
 
 prctl = ctypes.cdll.LoadLibrary('libc.so.6').prctl
 logger = logging.getLogger(__name__)
-RUN = os.environ['XDG_RUNTIME_DIR'] or '/run'
```
Comment (allisonkarlitskaya) [inline src/ferny.py:32]: lol.  oops.

File: test/test_basic.py [hunk 5]
```diff
@@ -34,7 +34,7 @@ async def test_rsa_key(self):
             await session.connect(server.host,
                                   port=server.port,
                                   configfile='none',
-                                  options={'UserKnownHostsFile': '/dev/null'},
+                                  options={'UserKnownHostsFile': '/dev/null', 'StrictHostKeyChecking': 'no'},
```
Comment (allisonkarlitskaya) [inline test/test_basic.py:37]: I wonder if those deserve to be on separate lines (with an extra indent) at this point...

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
:rofl: 

Thanks for all of these.  See comments below.

File: pyproject.toml (hunk 2 above)
Comment (martinpitt) [reply pyproject.toml:10]: So you object to the declaration of test dependencies itself? (See https://peps.python.org/pep-0631/#optional-dependencies ). Or only to using pip to install them? For Fedora etc. we can use `dnf`, but for CentOS and Debian (where mock-ssh is not packaged), we'll have to call pip, I'm afraid.

File: .flake8 [hunk 6]
```diff
@@ -0,0 +1,6 @@
+[flake8]
+ignore=
+    # line break after binary operator
+    W504
+
+max-line-length = 120
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply .flake8:6]: Done.

File: pyproject.toml (hunk 1 above)
Comment (martinpitt) [reply pyproject.toml:15]: No, just a typo. Fixed.

File: pyproject.toml (hunk 2 above)
Comment (martinpitt) [reply pyproject.toml:10]: Dropped the commit, as per IRC.

File: test/test_basic.py (hunk 5 above)
Comment (martinpitt) [reply test/test_basic.py]: It's "just" 109 chars on that line, but I don't mind. It looks nicer as separate lines. Done.

File: test/test_basic.py (hunk 3 above)
Comment (martinpitt) [reply test/test_basic.py]: It makes interactive debugging a bit more awkward, but then one can still chmod the original files. Changed to a temp dir.

**Comment by martinpitt** [comment]
@allisonkarlitskaya : Tests now work fine in GH actions, I updated the description.

**Review by allisonkarlitskaya — APPROVED** [review]
Thanks!

**Pull request merged**